packages/react-meteor-data/useTracker.js

```diff
@@ -28,41 +28,101 @@ function checkCursor(data) {
+  );
+}
+
+// inspired by https://github.com/facebook/react/blob/
+// 34ce57ae751e0952fd12ab532a3e5694445897ea/packages/
+// react-reconciler/src/ReactFiberHooks.js#L307-L354
+// used to replicate dep change behavior and stay consistent
+// with React.useEffect()
+function areHookInputsEqual(nextDeps, prevDeps) {
+  if (!nextDeps || !prevDeps) {
+    return false;
+  }
+
+  const len = nextDeps.length;
+
+  if (prevDeps.length !== len) {
+    return false;
+  }
+
+  for (let i = 0; i < len; i++) {
+    if (!is(nextDeps[i], prevDeps[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function useTracker(reactiveFn, deps) {
+  const previousDeps = useRef();
+  const computation = useRef();
+  const trackerData = useRef();
+
+  const [, forceUpdate] = useState();
+
+  const dispose = () => {
+    if (computation.current) {
+      computation.current.stop();
+      computation.current = null;
+    }
+  };
+
+  // this is called like at componentWillMount and componentWillUpdate equally
+  // in order to support render calls with synchronous data from the reactive computation
+  // if prevDeps or deps are not set areHookInputsEqual always returns false
+  // and the reactive functions is always called
+  if (!areHookInputsEqual(deps, previousDeps.current)) {
+    dispose();
+
+    // Use Tracker.nonreactive in case we are inside a Tracker Computation.
+    // This can happen if someone calls `ReactDOM.render` inside a Computation.
+    // In that case, we want to opt out of the normal behavior of nested
+    // Computations, where if the outer one is invalidated or stopped,
+    // it stops the inner one.
+    computation.current = Tracker.nonreactive(() => (
+      Tracker.autorun((c) => {
+        if (c.firstRun) {
+          const data = reactiveFn();
+          Meteor.isDevelopment && checkCursor(data);
+
+          // store the deps for comparison on next render
+          previousDeps.current = deps;
+          trackerData.current = data;
+        } else {
+          // makes sure that shallowEqualArray returns false
+          // which is always the case when prev or nextDeps are not set
+          previousDeps.current = null;
+          // Stop this computation instead of using the re-run.
+          // We use a brand-new autorun for each call to getMeteorData
+          // to capture dependencies on any reactive data sources that
+          // are accessed.  The reason we can't use a single autorun
+          // for the lifetime of the component is that Tracker only
+          // re-runs autoruns at flush time, while we need to be able to
+          // re-call the reactive function synchronously whenever we want, e.g.
+          // from next render.
+          c.stop();
+          // use Math.random() to trigger a state change to enforce a re-render
+          // Calling forceUpdate() triggers componentWillUpdate which
+          // calls the reactive function and re-renders the component.
+          forceUpdate(Math.random());
```

Comment:
The note I made about the use of `Math.random` was actually meant for here - it's super unlikely, but possible that `Math.random()` produces the same value as is already stored, which would mean it wouldn't trigger update. An incrementing value would work just as well here, and would prevent that from happening.

```js
const [forceValue, forceUpdate] = useState();
...
forceUpdate(forceValue++);
```

If there is something about react concurrent that might throw a wrench in the gears for that, then please disregard.

Reply:
Thanks for the catch, I totally missed that. A simple counter is good enough, https://github.com/yched/react-packages/pull/1/commits/c3803b96cf2eb843c8df549753586d398352f201 should do the job. (initially I wanted to use the tracker._id with is a counter itself, but since it is a private variable I used an own counter)